keps/sig-cluster-lifecycle/kubeadm/4656-add-instance-configuration/README.md

```diff
@@ -0,0 +1,382 @@
+
+* `kubeadm upgrade apply` will read the CRI annotaion and patch it to the global configuration and generate `/var/lib/kubelet/instace-config.yaml`, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.
+
+kubeadm upgrade node:
+
+* for other nodes, `kubeadm upgrade node` will check if the CRI annotation and  generate `/var/lib/kubelet/instace-config.yaml` based on them, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.
+
+**Beta(v1.32)**:
+
+* `kubeadm upgrade apply` and `kubeadm upgrade node`  will use `/var/lib/kubelet/instace-config.yaml`.
+
+**GA(v1.34)**:
+
+* `kubeadm upgrade apply` and `kubeadm upgrade node`  will use `/var/lib/kubelet/instace-config.yaml` and remove CRI annotation from Node object.
+
+### Test Plan
+
+[x ] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+At least the following kubeadm packages will require updates and new unit tests:
+
+* `cmd/kubeadm/app/cmd/phases/init`
+
+* `cmd/kubeadm/app/phases/kubelet`
+
+* `cmd/kubeadm/app/phases/upgrade`
+
+* `cmd/kubeadm/app/cmd/phases/join`
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+* A new e2e test will be added by using the kinder tool.
+
+### Graduation Criteria
+
+#### Alpha
+
+- The upgrade will still respect the CRI annotation.
+- Use `NodeLocalCRISocket` feature gate to implement features.
+- Add corresponding e2e tests.
+
+#### Beta
+
+* Make feature gate to be enabled by default.
+
+- Use instance kubelet configuration, ignore CRI annotations.
+- Gather feedback from developers and surveys.
+
+#### GA
+
+- Remove the CRI annotation during upgrade(this may be not urgent or have to).
+
+### Upgrade / Downgrade Strategy
+
+**Alpha**: Users can patch their `ClusterConfiguration` in the `kube-system/kubeadm-config` ConfigMap to enable the `NodeLocalCRISocket` feature gate before calling kubeadm upgrade apply, which will allow a `/var/lib/kubelet/instace-config.yaml` to be generated and overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` with it.
+
+**Beta**: Users can modify `ClusterConfiguration` to disable the feature gate during upgrades. This will allow them to continue using the CRI socket annotation on nodes.
+
+**GA**: Users can no longer patch ClusterConfiguration to opt out of the feature and it will be locked to be enabled by default.
+
+### Version Skew Strategy
+
+kubeadm will continue to skew from kubelet for three versions. The `ContainerRuntimeEndpoint` field in `KubeletConfiguration` was [introduced](https://github.com/kubernetes/kubernetes/pull/112136) in v1.27, so when we overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` through `/var/lib/kubelet/instace-config.yaml`, it will not have any impact on kubelet.
```

Comment:
it will not have any impact on kubelet. -> it will be supported on all versions of the kubelet within the skew.